File: bindings/node/__tests__/helpers/ws_server.js
'use strict';

const crypto = require('crypto');
const net = require('net');

const WS_GUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11';

function acceptKey(key) {
  return crypto.createHash('sha1').update(`${key}${WS_GUID}`).digest('base64');
}

function parseHeaders(raw) {
  const [requestLine, ...lines] = raw.split('\r\n');
  const headers = {};

  for (const line of lines) {
    if (!line) continue;
    const index = line.indexOf(':');
    if (index === -1) continue;

    const name = line.slice(0, index).trim().toLowerCase();
    const value = line.slice(index + 1).trim();
    if (headers[name] === undefined) {
      headers[name] = value;
    } else if (Array.isArray(headers[name])) {
      headers[name].push(value);
    } else {
      headers[name] = [headers[name], value];
    }
  }

  return { requestLine, headers };
}

function encodeFrame(opcode, payload = Buffer.alloc(0)) {
  const body = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
  let header;

  if (body.length < 126) {
    header = Buffer.from([0x80 | opcode, body.length]);
  } else if (body.length <= 0xffff) {
    header = Buffer.alloc(4);
    header[0] = 0x80 | opcode;
    header[1] = 126;
    header.writeUInt16BE(body.length, 2);
  } else {
    header = Buffer.alloc(10);
    header[0] = 0x80 | opcode;
    header[1] = 127;
    header.writeBigUInt64BE(BigInt(body.length), 2);
  }

  return Buffer.concat([header, body]);
}

function decodeFrame(buffer) {
  if (buffer.length < 2) return null;

  const first = buffer[0];
  const second = buffer[1];
  const masked = (second & 0x80) !== 0;
  let length = second & 0x7f;
  let offset = 2;

  if (length === 126) {
    if (buffer.length < offset + 2) return null;
    length = buffer.readUInt16BE(offset);
    offset += 2;
  } else if (length === 127) {
    if (buffer.length < offset + 8) return null;
    const largeLength = buffer.readBigUInt64BE(offset);
    if (largeLength > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new Error('WebSocket frame too large for test helper');
    }
    length = Number(largeLength);
    offset += 8;
  }

  const maskOffset = offset;
  if (masked) offset += 4;
  if (buffer.length < offset + length) return null;

  const payload = Buffer.from(buffer.subarray(offset, offset + length));
  if (masked) {
    const mask = buffer.subarray(maskOffset, maskOffset + 4);
    for (let index = 0; index < payload.length; index += 1) {
      payload[index] ^= mask[index % 4];
    }
  }

  return {
    frame: {
      fin: (first & 0x80) !== 0,
      opcode: first & 0x0f,
      payload,
    },
    rest: buffer.subarray(offset + length),
  };
}

function chooseProtocol(requested, supported) {
  if (!requested || supported.length === 0) return undefined;
  const candidates = requested.split(',').map((value) => value.trim()).filter(Boolean);
  return candidates.find((value) => supported.includes(value));
}

function createWebSocketFixture(options = {}) {
  const host = options.host || '127.0.0.1';
  const supportedProtocols = options.protocols || [];
  const connections = [];
  const handshakes = [];
  const server = net.createServer();

  server.on('connection', (socket) => {
    let handshakeBuffer = Buffer.alloc(0);
    let frameBuffer = Buffer.alloc(0);
    let connected = false;
    let selectedProtocol;

    socket.on('data', (chunk) => {
      if (!connected) {
        handshakeBuffer = Buffer.concat([handshakeBuffer, chunk]);
        const end = handshakeBuffer.indexOf('\r\n\r\n');
        if (end === -1) return;

        const head = handshakeBuffer.subarray(0, end).toString('latin1');
        frameBuffer = handshakeBuffer.subarray(end + 4);
        const { requestLine, headers } = parseHeaders(head);
        selectedProtocol = chooseProtocol(headers['sec-websocket-protocol'], supportedProtocols);
        handshakes.push({ requestLine, headers, selectedProtocol });

        const response = [
          'HTTP/1.1 101 Switching Protocols',
          'Upgrade: websocket',
          'Connection: Upgrade',
          `Sec-WebSocket-Accept: ${acceptKey(headers['sec-websocket-key'] || '')}`,
          selectedProtocol ? `Sec-WebSocket-Protocol: ${selectedProtocol}` : undefined,
          ...(options.setCookie ? [`Set-Cookie: ${options.setCookie}`] : []),
          '\r\n',
        ].filter(Boolean).join('\r\n');

        socket.write(response);
        connected = true;
      } else {
        frameBuffer = Buffer.concat([frameBuffer, chunk]);
      }

      while (connected) {
        const decoded = decodeFrame(frameBuffer);
        if (!decoded) break;
        frameBuffer = decoded.rest;
        const { opcode, payload } = decoded.frame;

        if (opcode === 0x1 || opcode === 0x2) {
          socket.write(encodeFrame(opcode, payload));
        } else if (opcode === 0x9) {
          socket.write(encodeFrame(0xA, payload));
        } else if (opcode === 0x8) {
          socket.write(encodeFrame(0x8, payload));
          socket.end();
          break;
        }
      }
    });

    connections.push({
      socket,
      get selectedProtocol() {
        return selectedProtocol;
      },
      sendText(text) {
        socket.write(encodeFrame(0x1, Buffer.from(text)));
      },
      sendBinary(data) {
        socket.write(encodeFrame(0x2, Buffer.from(data)));
      },
      ping(data = Buffer.alloc(0)) {
        socket.write(encodeFrame(0x9, Buffer.from(data)));
      },
      close(code = 1000, reason = '') {
        const payload = Buffer.alloc(2 + Buffer.byteLength(reason));
        payload.writeUInt16BE(code, 0);
        payload.write(reason, 2);
        socket.write(encodeFrame(0x8, payload));
      },
    });
  });

  return {
    handshakes,
    connections,
    async start() {
      await new Promise((resolve) => server.listen(0, host, resolve));
      const address = server.address();
      return {
        host: address.address,
        port: address.port,
        url: `ws://${address.address}:${address.port}/ws`,
      };
    },
    async stop() {
      for (const connection of connections) {
        connection.socket.destroy();
      }
      await new Promise((resolve, reject) => {
        server.close((error) => (error ? reject(error) : resolve()));
      });
    },
  };
}

module.exports = {
  createWebSocketFixture,
  encodeFrame,
  decodeFrame,
};

File: bindings/python/tests/helpers/ws_server.py
"""Local deterministic RFC 6455 WebSocket fixture for binding tests."""

from __future__ import annotations

import asyncio
import base64
import hashlib
import struct
from dataclasses import dataclass, field
from typing import Callable


WS_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"


def accept_key(key: str) -> str:
    digest = hashlib.sha1(f"{key}{WS_GUID}".encode("ascii")).digest()
    return base64.b64encode(digest).decode("ascii")


def encode_frame(opcode: int, payload: bytes = b"") -> bytes:
    length = len(payload)
    if length < 126:
        header = bytes([0x80 | opcode, length])
    elif length <= 0xFFFF:
        header = bytes([0x80 | opcode, 126]) + struct.pack("!H", length)
    else:
        header = bytes([0x80 | opcode, 127]) + struct.pack("!Q", length)
    return header + payload


async def read_frame(reader: asyncio.StreamReader) -> tuple[int, bytes]:
    first, second = await reader.readexactly(2)
    masked = bool(second & 0x80)
    length = second & 0x7F

    if length == 126:
        length = struct.unpack("!H", await reader.readexactly(2))[0]
    elif length == 127:
        length = struct.unpack("!Q", await reader.readexactly(8))[0]

    mask = await reader.readexactly(4) if masked else b""
    payload = bytearray(await reader.readexactly(length))
    if masked:
        for index, value in enumerate(payload):
            payload[index] = value ^ mask[index % 4]

    return first & 0x0F, bytes(payload)


def parse_headers(raw: bytes) -> tuple[str, dict[str, str]]:
    lines = raw.decode("latin1").split("\r\n")
    request_line = lines[0]
    headers: dict[str, str] = {}

    for line in lines[1:]:
        if not line or ":" not in line:
            continue
        name, value = line.split(":", 1)
        headers[name.strip().lower()] = value.strip()

    return request_line, headers


@dataclass
class WebSocketHandshake:
    request_line: str
    headers: dict[str, str]
    selected_protocol: str | None


@dataclass
class WebSocketConnection:
    reader: asyncio.StreamReader
    writer: asyncio.StreamWriter
    selected_protocol: str | None
    received: list[tuple[str, bytes]] = field(default_factory=list)

    async def send_text(self, text: str) -> None:
        self.writer.write(encode_frame(0x1, text.encode()))
        await self.writer.drain()

    async def send_binary(self, data: bytes) -> None:
        self.writer.write(encode_frame(0x2, data))
        await self.writer.drain()

    async def ping(self, data: bytes = b"") -> None:
        self.writer.write(encode_frame(0x9, data))
        await self.writer.drain()

    async def close(self, code: int = 1000, reason: str = "") -> None:
        payload = struct.pack("!H", code) + reason.encode()
        self.writer.write(encode_frame(0x8, payload))
        await self.writer.drain()
        await close_writer(self.writer)


class WebSocketFixture:
    def __init__(
        self,
        host: str = "127.0.0.1",
        protocols: list[str] | None = None,
        set_cookie: str | None = None,
        on_connection: Callable[[WebSocketConnection], None] | None = None,
    ) -> None:
        self.host = host
        self.protocols = protocols or []
        self.set_cookie = set_cookie
        self.on_connection = on_connection
        self.handshakes: list[WebSocketHandshake] = []
        self.connections: list[WebSocketConnection] = []
        self.server: asyncio.AbstractServer | None = None
        self.port: int | None = None
        self.url: str | None = None

    async def __aenter__(self) -> "WebSocketFixture":
        await self.start()
        return self

    async def __aexit__(self, *_exc: object) -> None:
        await self.stop()

    async def start(self) -> "WebSocketFixture":
        self.server = await asyncio.start_server(self._handle, self.host, 0)
        socket = self.server.sockets[0]
        self.port = socket.getsockname()[1]
        self.url = f"ws://{self.host}:{self.port}/ws"
        return self

    async def stop(self) -> None:
        for connection in self.connections:
            connection.writer.transport.abort()
        if self.server is not None:
            self.server.close()
            await self.server.wait_closed()
            self.server = None

    async def _handle(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
    ) -> None:
        raw = await reader.readuntil(b"\r\n\r\n")
        request_line, headers = parse_headers(raw.rstrip(b"\r\n"))
        selected = self._select_protocol(headers.get("sec-websocket-protocol"))
        self.handshakes.append(WebSocketHandshake(request_line, headers, selected))

        response = [
            "HTTP/1.1 101 Switching Protocols",
            "Upgrade: websocket",
            "Connection: Upgrade",
            f"Sec-WebSocket-Accept: {accept_key(headers.get('sec-websocket-key', ''))}",
        ]
        if selected:
            response.append(f"Sec-WebSocket-Protocol: {selected}")
        if self.set_cookie:
            response.append(f"Set-Cookie: {self.set_cookie}")
        writer.write(("\r\n".join(response) + "\r\n\r\n").encode("ascii"))
        await writer.drain()

        connection = WebSocketConnection(reader, writer, selected)
        self.connections.append(connection)
        if self.on_connection:
            self.on_connection(connection)

        try:
            while not reader.at_eof():
                try:
                    opcode, payload = await read_frame(reader)
                except asyncio.IncompleteReadError:
                    break

                if opcode == 0x1:
                    connection.received.append(("text", payload))
                    writer.write(encode_frame(0x1, payload))
                elif opcode == 0x2:
                    connection.received.append(("binary", payload))
                    writer.write(encode_frame(0x2, payload))
                elif opcode == 0x9:
                    connection.received.append(("ping", payload))
                    writer.write(encode_frame(0xA, payload))
                elif opcode == 0x8:
                    connection.received.append(("close", payload))
                    writer.write(encode_frame(0x8, payload))
                    await writer.drain()
                    return
                await writer.drain()
        finally:
            await close_writer(writer)

    def _select_protocol(self, requested: str | None) -> str | None:
        if not requested:
            return None
        candidates = [value.strip() for value in requested.split(",")]
        return next((value for value in candidates if value in self.protocols), None)


def create_websocket_fixture(**kwargs: object) -> WebSocketFixture:
    return WebSocketFixture(**kwargs)


async def close_writer(writer: asyncio.StreamWriter) -> None:
    writer.close()
    try:
        await asyncio.wait_for(writer.wait_closed(), timeout=1)
    except (asyncio.TimeoutError, OSError):
        writer.transport.abort()
